doc/manual/contract.md

```diff
@@ -0,0 +1,421 @@
+
+All contracts (`expect`, `ensure`) are evaluated in the same context as the called method, it's possible to refer to attributes, methods and parameters in the condition.
+
+When defining a contract (`expect`, `ensure`) on an attribute like bellow, only the automatically generated write property (setter) will use the contract.
+
+~~~
+	var baz: Int is expect(baz > 0) # In the contract, baz refers to the parameter of the write property (setter).
+~~~
+
+### Define preconditions with `expect`
+
+We define a precondition using the annotation `expect`. In the `Stack` class, the method `pop` defines a precondition, that a call to the method should be made on a non-empty stack.
+
+~~~
+	fun pop is expect(not_empty)
+~~~
+
+### Define postconditions with `ensure`
+
+We define a postcondition using the annotation `ensure`. In the `Stack` class, the method `push` defines postconditions. The first guarantee defines that, the method must add the element received as argument (`object`) at the top of the stack.
+
+~~~
+	fun push(object: T) is ensure(self[size - 1] == object)
+~~~
+
+Within an `ensure` contract we can refer to the return value of the method with the keyword `result`. In the `Stack` class, the `[]` method guarantees that, the return value is the same as the element stored in the stack at the given index.
+
+~~~
+	fun [](index: Int): T is ensure(result == data[index])
+~~~
+
+Note that the keyword `result` can only be used when the method has a return parameter.
+
+
+In `ensure` we can refer to any arguments and instance variables before the method call with the using of `old` keyword. Again, with the `pop` method, we can ensure that the `size` of the stack after the method call will be the same as before minus one `old(size) - 1 == size`.
+
+~~~
+	fun pop(object: T) is ensure(old(size) - 1 == size)
+~~~
+
+In an `old(expression)` the `expression` is evaluated in the same context as the called method. Thus, it's possible to refer to attributes, methods, and arguments.
+
+`old(object)` stores a reference to object. By doing so, if `object` is mutated in the method the `old (object)` will also be mutated. To avoid this, it's necessary to provide a way to duplicate the object like `clone` method so that `not old(object.clone).is_same_instance (object)` will be true. In the stack example, the expression `old(stack).is_same_instance(stack) && old(stack.size) != old(stack).size` is true. If we need to `old(stack).is_same_instance(stack)` to be false, we need to provide a method to duplicate the object, such as `not old(stack.clone).is_same_instance(stack)` be true.
```

Comment:
instead of dupicatin object is it often better to captue only the insteresting part. e.g.

```nit
ensure(old(foo).bar == foo.bar + 1) # bad 
ensure(old(foo.bar) == foo.bar + 1) # better
```